README.md
# Python Port Scanner v0.2

An easy to use Python package that could perform port scanning conveniently.

An output example is showed as following:
![Output Example](https://github.com/YaokaiYang-assaultmaster/PythonPortScanner/blob/master/ExampleGraph/ScanResultExample.png)

## Installation  
1. Use `git clone https://github.com/YaokaiYang-assaultmaster/PythonPortScanner.git` command to clone the repository to your own machine.   
2. There is 2 methods for you to use this module.   
	- One is to put `PortScanner package` in your working directory and import `PortScanner` to your code. In this way you can only use it in your current working directory.  
	- The other is to execute `python setup.py install` in the `PortScanner` package. In this way it will be installed to your python's `site-packages` folder. And after that you can use it globally by `import PortScanner`.   
3. Voilà! You are ready to go!

## Usage  
1. Add `import PortScanner` or `from PortScanner import PortScanner`in your code.  
2. Initialize a new PortScanner object using `scanner = PortScanner.PortScanner()` or `scanner = PortScanner()`. You could also put the list of ports you want to scan (if any) as a python `list` and pass it as the `target_ports` argument to the constructor.  
3. Then call `scanner.scan(host_name)` to perform scan task. 
4. __Note that the total scan time for a target website is highly related to the timeout value (delay) set for the Scanner object. Thus for the seek of efficiency, the timeout should not be too long.__

## Documentation 
- Constructor  
`__init__(self, target_ports=None)` takes a list of ports or an int as the argument. If not provided, it will perform scanning task using default ports. If a list is provided, the list will be used as the list of ports being scanned. If a int is provided (this int need to be 50, 100 or 1000), the top 50, 100 or 1000 commonly used ports will be used.     

- Functions  
	1. ` scanner.scan(host_name, message = '')` is the function need to be called to perform port scanning. It takes 2 arguments.   
		- `host_name` is the hostname that is going to be scanned
    	- `message` is the message that is going to be included in the scanning packets sent out. This is provided in order to prevent ethical problem. If not provided, no message will be included in the packets.  
	2.  `scanner.set_thread_limit(limit)` is the function to set the maximum number of threads run concurrently for port scanning. It takes 1 argument.  
		- `limit` is the maximum number of threads allowed. The valid limit range is 1 to 50,000. The default value is 1,000.   
	3.  `scanner.set_delay(delay)` is the function to set the timeout delay for port scanning in seconds. It takes 1 argument. 
		- `delay` the time in seconds that a TCP socket waits until timeout. The valid delay range is 1s to 100s. The default value is 10s.   
	4. `scanner.show_target_ports()` is used to get the list of ports being scanned for current Scanner object.     
	5. `scanner.show_delay()` is used to get current timeout interval in seconds that a TCP socket waits.       
	6. `scanner.show_top_k_ports(k)` is used to get top 50, top 100 or top 1000 port lists. Other k will raise an `ValueError` 

- __An example usage case is showed in `PortScanner/PortScanExample.py`__

## Change logs can be found [here](https://github.com/YaokaiYang-assaultmaster/PythonPortScanner/blob/master/CHANGELOG.md)


CHANGELOG.md
# Changelog

Versions are monotonically increased based on Semantic Versioning.

***

## V0.2 (09/12/2017)

## Backward incompatible changes
None

## Deprecations
None

## Changes
Fixed [issue #1](https://github.com/YaokaiYang-assaultmaster/PythonPortScanner/issues/1). Specifically, fixed Windows has no `SO_REUSEPORT` for TCP socket error and change message encoding to `utf-8` before sending out. 

***

## V0.1 (05/30/2017)

## Backward incompatible changes
None

## Deprecations
None

## Changes
Initialized the whole project. First version finished. 


PortScanner/etc/constants.py
"""
This files contains our three list of default ports to be scanned. They are top 50, top 100 and top 1000 ports
commonly used, respectively.
"""


port_list_top_1000 = [1, 3, 6, 9, 13, 17, 19, 20, 21, 22, 23, 24, 25, 30, 32, 37, 42, 49, 53, 70, 79, 80, 81, 82, 83,
                      84,
                      88, 89, 99, 106, 109, 110, 113, 119, 125, 135, 139, 143, 146, 161, 163, 179, 199, 211, 222, 254,
                      255,
                      259, 264, 280, 301, 306, 311, 340, 366, 389, 406, 416, 425, 427, 443, 444, 458, 464, 481, 497,
                      500,
                      512, 513, 514, 524, 541, 543, 544, 548, 554, 563, 587, 593, 616, 625, 631, 636, 646, 648, 666,
                      667,
                      683, 687, 691, 700, 705, 711, 714, 720, 722, 726, 749, 765, 777, 783, 787, 800, 808, 843, 873,
                      880,
                      888, 898, 900, 901, 902, 911, 981, 987, 990, 992, 995, 999, 1000, 1001, 1007, 1009, 1010, 1021,
                      1022,
                      1023, 1024, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1037, 1038,
                      1039,
                      1040, 1041, 1042, 1043, 1044, 1045, 1046, 1047, 1048, 1049, 1050, 1051, 1052, 1053, 1054, 1055,
                      1056,
                      1057, 1058, 1059, 1060, 1061, 1062, 1063, 1064, 1065, 1066, 1067, 1068, 1069, 1070, 1071, 1072,
                      1073,
                      1074, 1075, 1076, 1077, 1078, 1079, 1080, 1081, 1082, 1083, 1084, 1085, 1086, 1087, 1088, 1089,
                      1090,
                      1091, 1092, 1093, 1094, 1095, 1096, 1097, 1098, 1099, 1102, 1104, 1105, 1106, 1107, 1110, 1111,
                      1112,
                      1113, 1117, 1119, 1121, 1122, 1123, 1126, 1130, 1131, 1137, 1141, 1145, 1147, 1148, 1151, 1154,
                      1163,
                      1164, 1165, 1169, 1174, 1183, 1185, 1186, 1192, 1198, 1201, 1213, 1216, 1217, 1233, 1236, 1244,
                      1247,
                      1259, 1271, 1277, 1287, 1296, 1300, 1309, 1310, 1322, 1328, 1334, 1352, 1417, 1433, 1443, 1455,
                      1461,
                      1494, 1500, 1503, 1521, 1524, 1533, 1556, 1580, 1583, 1594, 1600, 1641, 1658, 1666, 1687, 1700,
                      1717,
                      1718, 1719, 1720, 1723, 1755, 1761, 1782, 1801, 1805, 1812, 1839, 1862, 1863, 1875, 1900, 1914,
                      1935,
                      1947, 1971, 1974, 1984, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
                      2013,
                      2020, 2021, 2030, 2033, 2034, 2038, 2040, 2041, 2042, 2045, 2046, 2047, 2048, 2065, 2068, 2099,
                      2103,
                      2105, 2106, 2111, 2119, 2121, 2126, 2135, 2144, 2160, 2170, 2179, 2190, 2196, 2200, 2222, 2251,
                      2260,
                      2288, 2301, 2323, 2366, 2381, 2382, 2393, 2399, 2401, 2492, 2500, 2522, 2525, 2557, 2601, 2604,
                      2607,
                      2638, 2701, 2710, 2717, 2725, 2800, 2809, 2811, 2869, 2875, 2909, 2920, 2967, 2998, 3000, 3003,
                      3005,
                      3006, 3011, 3013, 3017, 3030, 3052, 3071, 3077, 3128, 3168, 3211, 3221, 3260, 3268, 3283, 3300,
                      3306,
                      3322, 3323, 3324, 3333, 3351, 3367, 3369, 3370, 3371, 3389, 3404, 3476, 3493, 3517, 3527, 3546,
                      3551,
                      3580, 3659, 3689, 3703, 3737, 3766, 3784, 3800, 3809, 3814, 3826, 3827, 3851, 3869, 3871, 3878,
                      3880,
                      3889, 3905, 3914, 3918, 3920, 3945, 3971, 3986, 3995, 3998, 4000, 4001, 4002, 4003, 4004, 4005,
                      4045,
                      4111, 4125, 4129, 4224, 4242, 4279, 4321, 4343, 4443, 4444, 4445, 4449, 4550, 4567, 4662, 4848,
                      4899,
                      4998, 5000, 5001, 5002, 5003, 5009, 5030, 5033, 5050, 5054, 5060, 5080, 5087, 5100, 5101, 5120,
                      5190,
                      5200, 5214, 5221, 5225, 5269, 5280, 5298, 5357, 5405, 5414, 5431, 5440, 5500, 5510, 5544, 5550,
                      5555,
                      5560, 5566, 5631, 5633, 5666, 5678, 5718, 5730, 5800, 5801, 5810, 5815, 5822, 5825, 5850, 5859,
                      5862,
                      5877, 5900, 5901, 5902, 5903, 5906, 5910, 5915, 5922, 5925, 5950, 5952, 5959, 5960, 5961, 5962,
                      5987,
                      5988, 5998, 5999, 6000, 6001, 6002, 6003, 6004, 6005, 6006, 6009, 6025, 6059, 6100, 6106, 6112,
                      6123,
                      6129, 6156, 6346, 6389, 6502, 6510, 6543, 6547, 6565, 6566, 6580, 6646, 6666, 6667, 6668, 6689,
                      6692,
                      6699, 6779, 6788, 6792, 6839, 6881, 6901, 6969, 7000, 7001, 7004, 7007, 7019, 7025, 7070, 7100,
                      7103,
                      7106, 7200, 7402, 7435, 7443, 7496, 7512, 7625, 7627, 7676, 7741, 7777, 7800, 7911, 7920, 7937,
                      7999,
                      8000, 8001, 8007, 8008, 8009, 8010, 8021, 8031, 8042, 8045, 8080, 8081, 8082, 8083, 8084, 8085,
                      8086,
                      8087, 8088, 8089, 8093, 8099, 8180, 8192, 8193, 8200, 8222, 8254, 8290, 8291, 8300, 8333, 8383,
                      8400,
                      8402, 8443, 8500, 8600, 8649, 8651, 8654, 8701, 8800, 8873, 8888, 8899, 8994, 9000, 9001, 9002,
                      9009,
                      9010, 9040, 9050, 9071, 9080, 9090, 9099, 9100, 9101, 9102, 9110, 9200, 9207, 9220, 9290, 9415,
                      9418,
                      9485, 9500, 9502, 9535, 9575, 9593, 9594, 9618, 9666, 9876, 9877, 9898, 9900, 9917, 9929, 9943,
                      9968,
                      9998, 9999, 10000, 10001, 10002, 10003, 10009, 10012, 10024, 10082, 10180, 10215, 10243, 10566,
                      10616, 10621, 10626, 10628, 10778, 11110, 11967, 12000, 12174, 12265, 12345, 13456, 13722,
                      13782,
                      14000, 14238, 14441, 15000, 15002, 15003, 15660, 15742, 16000, 16012, 16016, 16018, 16080,
                      16113,
                      16992, 17877, 17988, 18040, 18101, 18988, 19101, 19283, 19315, 19350, 19780, 19801, 19842,
                      20000,
                      20005, 20031, 20221, 20828, 21571, 22939, 23502, 24444, 24800, 25734, 26214, 27000, 27352,
                      27355,
                      27715, 28201, 30000, 30718, 30951, 31038, 31337, 32768, 32769, 32770, 32771, 32772, 32773,
                      32774,
                      32775, 32776, 32777, 32778, 32779, 32780, 32781, 32782, 32783, 32784, 33354, 33899, 34571,
                      34572,
                      35500, 38292, 40193, 40911, 41511, 42510, 44176, 44442, 44501, 45100, 48080, 49152, 49153,
                      49154,
                      49155, 49156, 49157, 49158, 49159, 49160, 49163, 49165, 49167, 49175, 49400, 49999, 50000,
                      50001,
                      50002, 50006, 50300, 50389, 50500, 50636, 50800, 51103, 51493, 52673, 52822, 52848, 52869,
                      54045,
                      54328, 55055, 55555, 55600, 56737, 57294, 57797, 58080, 60020, 60443, 61532, 61900, 62078,
                      63331,
                      64623, 64680, 65000, 65129, 65389]
port_list_top_100 = [7, 9, 13, 21, 22, 25, 37, 53, 79, 80, 88, 106, 110, 113, 119, 135, 139, 143, 179, 199, 389,
                     427, 443, 444, 465, 513, 514, 543, 548, 554, 587, 631, 646, 873, 990, 993, 995, 1025, 1026,
                     1027, 1028, 1110, 1433, 1720, 1723, 1755, 1900, 2000, 2049, 2121, 2717, 3000, 3128, 3306,
                     3389, 3986, 4899, 5000, 5009, 5051, 5060, 5101, 5190, 5357, 5432, 5631, 5666, 5800, 5900,
                     6000, 6646, 7070, 8000, 8008, 8080, 8443, 8888, 9100, 9999, 32768, 49152, 49153, 49154,
                     49155, 49156]
port_list_top_50 = [21, 22, 25, 53, 80, 110, 113, 135, 139, 143, 179, 199, 443, 445, 465, 514, 548, 554, 587, 646,
                    993, 995, 1025, 1026, 1433, 1720, 1723, 2000, 3306, 3389, 5060, 5666, 5900, 6001, 8000, 8008,
                    8080, 8443, 8888, 10000, 32768, 49152, 49154]


PortScanner/PortScanExample.py
import PortScanner as ps


def main():
    # Initialize a Scanner object that will scan top 50 commonly used ports.
    scanner = ps.PortScanner(target_ports=50)

    host_name = 'google.com'

    message = 'put whatever message you want here'

    '''
    output contains a dictionary of {port:status} pairs
    in which port is the list of ports we scanned 
    and status is either 'OPEN' or 'CLOSE'
    '''

    # This line sets the thread limit of the scanner to 1500
    scanner.set_thread_limit(1500)

    # This line sets the timeout delay to 15s
    scanner.set_delay(15)

    # This line shows the target port list of the scanner
    scanner.show_target_ports()
    '''
    Current port list is: 
    [blah, blah ....]
    '''

    # This line shows the timeout delay of the scanner
    scanner.show_delay()
    '''
    Current timeout delay is 15 seconds.
    '''

    # This line shows the top 100 commonly used ports.
    scanner.show_top_k_ports(100)
    '''
    Top 100 commonly used ports:
    [blah, blah ....]
    '''

    output = scanner.scan(host_name, message)
    '''
    start scanning website: google.com
    server ip is: 172.217.4.110
    80: OPEN

    443: OPEN

    2000: OPEN

    5060: OPEN

    host google.com scanned in  30.956103 seconds
    finish scanning!
    '''


if __name__ == "__main__":
    main()


PortScanner/PortScanner.py
# -*- coding: utf-8 -*-

import socket
import platform
import threading
import time

from etc import constants


class PortScanner:
    # default ports to be scanned is top 1000
    __port_list_top_1000 = constants.port_list_top_1000
    __port_list_top_100 = constants.port_list_top_100
    __port_list_top_50 = constants.port_list_top_50

    # default thread number limit
    __thread_limit = 1000

    # default connection timeout time in seconds
    __delay = 10


    @classmethod
    def __usage(cls):
        """
        Return the usage information for invalid input host name.
        """
        print('python Port Scanner v0.1')
        print('please make sure the input host name is in the form of "something.com" or "http://something.com!"\n')

    def __init__(self, target_ports=None):
        """
        Constructor of a PortScanner object. If target_ports is a list, this list of ports will be used as
        the port list to be scanned. If the target_ports is a int, it should be 50, 100 or 1000, indicating
        which default list will be used.

        :param target_ports: if this args is a list, then this list of ports that is going to be scanned,
        default to self.__port_list_top_1000. if this args is an int, then it should be 50, 100 or 1000. And
        the corresponding default list will be used respectively.
        :type target_ports: list or int
        """
        if target_ports is None:
            self.target_ports = self.__port_list_top_1000
        elif type(target_ports) == list:
            self.target_ports = target_ports
        elif type(target_ports) == int:
            self.target_ports = self.check_default_list(target_ports)

    def check_default_list(self, target_port_rank):
        """
        Check the input target port rank. The target port rank should be 50, 100 or 1000.
        And for a valid input, corresponding port list will be returned.

        :param target_port_rank: top K commonly used port list to be returned.
        :return: top K commonly used port list.
        """
        if (
            target_port_rank != 50 and
            target_port_rank != 100 and
            target_port_rank != 1000
        ):
            raise ValueError(
                'Invalid port rank {}. Should be 50, 100 or 1,000.'.format(target_port_rank)
            )

        if target_port_rank == 50:
            return self.__port_list_top_50
        elif target_port_rank == 100:
            return self.__port_list_top_100
        else:
            return self.__port_list_top_1000

    def scan(self, host_name, message=''):
        """
        This is the function need to be called to perform port scanning

        :param host_name: the hostname that is going to be scanned
        :param message: the message that is going to be included in the scanning packets
        in order to prevent ethical problem (default: '').
        :return: a dict object containing the scan results for a given host in the form of
        {port_number: status}
        :rtype: dict
        """
        host_name = str(host_name)
        if 'http://' in host_name or 'https://' in host_name:
            host_name = host_name[host_name.find('://') + 3:]

        print('*' * 60 + '\n')
        print('start scanning website: {}'.format(host_name))

        try:
            server_ip = socket.gethostbyname(host_name)
            print('server ip is: {}'.format(str(server_ip)))

        except socket.error:
            # If the DNS resolution of a website cannot be finished, abort that website.
            print('hostname {} unknown!!!'.format(host_name))
            self.__usage()
            return {}
            # May need to return specific value to indicate the failure.

        start_time = time.time()
        output = self.__scan_ports(server_ip, self.__delay, message.encode('utf-8'))
        stop_time = time.time()

        print('host {} scanned in  {} seconds'.format(host_name, stop_time - start_time))
        print('finished scan!\n')

        return output

    def set_thread_limit(self, limit):
        """
        Set the maximum number of thread for port scanning

        :param limit: the maximum number of thread running concurrently, default to 1000.
        """
        limit = int(limit)

        if limit <= 0 or limit > 50000:
            print(
                'Warning: Invalid thread number limit {}!'
                'Please make sure the thread limit is within the range of (1, 50,000)!'.format(limit)
            )
            print('The scanning process will use default thread limit 1,000.')
            return

        self.__thread_limit = limit

    def set_delay(self, delay):
        """
        Set the time out delay for port scanning in seconds

        :param delay: the time in seconds that a TCP socket waits until timeout, default to 10s.
        """
        delay = int(delay)
        if delay <= 0 or delay > 100:
            print(
                'Warning: Invalid delay value {} seconds!'
                'Please make sure the input delay is within the range of (1, 100)'.format(delay)
            )
            print('The scanning process will use the default delay time 10 seconds.')
            return

        self.__delay = delay

    def show_target_ports(self):
        """
        Print out and return the list of ports being scanned.

        :return: list of ports scanned by current Scanner object.
        :rtype: list
        """
        print ('Current port list is:')
        print (self.target_ports)
        return self.target_ports

    def show_delay(self):
        """
        Print out and return the delay in seconds that a TCP socket waits until timeout.

        :return: timeout interval of the TCP connection in seconds.
        :rtype: int
        """
        print ('Current timeout delay is {} seconds.'.format(self.__delay))
        return self.__delay

    def show_top_k_ports(self, k):
        """
        Print out and return top K commonly used ports. K should be 50, 100 or 1000.

        :param k: top K list will be returned.
        :type k: int
        :return: top K commonly used ports.
        :rtype: list
        """
        port_list = self.check_default_list(k)
        print('Top {} commonly used ports:'.format(k))
        print(port_list)
        return port_list

    def __scan_ports_helper(self, ip, delay, output, message):
        """
        Open multiple threads to perform port scanning

        :param ip: the ip address that is being scanned
        :type ip: str
        :param delay: the time in seconds that a TCP socket waits until timeout
        :type delay: int
        :param output: a dict that stores result in {port, status} style pairs.
        status can be 'OPEN' or 'CLOSE'.
        :type output: dict
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        """
        port_index = 0

        while port_index < len(self.target_ports):
            # Ensure the number of concurrently running threads does not exceed the thread limit
            while threading.activeCount() < self.__thread_limit and port_index < len(self.target_ports):
                # Start threads
                thread = threading.Thread(target=self.__TCP_connect,
                                          args=(ip, self.target_ports[port_index], delay, output, message))
                thread.start()
                port_index = port_index + 1

    def __scan_ports(self, ip, delay, message):
        """
        Controller of the __scan_ports_helper() function

        :param ip: the ip address that is being scanned
        :type ip: str
        :param delay: the time in seconds that a TCP socket waits until timeout
        :type delay: int
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        :return: a dict that stores result in {port, status} style pairs.
        status can be 'OPEN' or 'CLOSE'.
        """
        output = {}

        thread = threading.Thread(target=self.__scan_ports_helper, args=(ip, delay, output, message))
        thread.start()

        # Wait until all ports being scanned
        while len(output) < len(self.target_ports):
            continue

        # Print opening ports from small to large
        for port in self.target_ports:
            if output[port] == 'OPEN':
                print('{}: {}\n'.format(port, output[port]))

        return output

    def __TCP_connect(self, ip, port_number, delay, output, message):
        """
        Perform status checking for a given port on a given ip address using TCP handshake

        :param ip: the ip address that is being scanned
        :type ip: str
        :param port_number: the port that is going to be checked
        :type port_number: int
        :param delay: the time in seconds that a TCP socket waits until timeout
        :type delay: int
        :param output: a dict that stores result in {port, status} style pairs.
        status can be 'OPEN' or 'CLOSE'.
        :type output: dict
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        """
        # Initialize the TCP socket object based on different operating systems.
        # All systems except for 'Windows' will be treated equally.
        curr_os = platform.system()
        if curr_os == 'Windows':
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            TCP_sock.settimeout(delay)
        else:
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
            TCP_sock.settimeout(delay)

        # Initialize a UDP socket to send scanning alert message if there exists an non-empty message
        if message != '':
            UDP_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            UDP_sock.sendto(message, (ip, int(port_number)))

        try:
            result = TCP_sock.connect_ex((ip, int(port_number)))
            if message != '':
                TCP_sock.sendall(message)

            # If the TCP handshake is successful, the port is OPEN. Otherwise it is CLOSE
            if result == 0:
                output[port_number] = 'OPEN'
            else:
                output[port_number] = 'CLOSE'

            TCP_sock.close()

        except socket.error as e:
            # Failed to perform a TCP handshake means the port is probably close.
            output[port_number] = 'CLOSE'
            pass
